ETTrace/Tracer/EMGStackTraceRecorder.cpp

```diff
@@ -0,0 +1,105 @@
+#include "EMGStackTraceRecorder.h"
+
+#import <QuartzCore/QuartzCore.h>
+#import <mach-o/arch.h>
+#import <mach/mach.h>
+#import <pthread.h>
+#import <deque>
+#import <iostream>
+#import <mutex>
+#import <unordered_map>
+
+extern "C" {
+void FIRCLSWriteThreadStack(thread_t thread, uintptr_t *frames, uint64_t framesCapacity, uint64_t *framesWritten);
+}
+
+static const int kMaxFramesPerStack = 1024;
+
+kern_return_t checkMachCall(kern_return_t result) {
+    if (result != KERN_SUCCESS) {
+        std::cerr << "Mach call failed with " << result << std::endl;
+    }
+    return result;
+}
+
+Thread::Thread(thread_t threadId, thread_t mainThreadId) {
+    name = "Failed to get name"; // Error case
+
+    if(threadId == mainThreadId) {
+        name = "Main Thread";
+    } else {
+        // Get thread Name
+        char cName[1024];
+        pthread_t pt = pthread_from_mach_thread_np(threadId);
+        if (pt) {
+            int rc = pthread_getname_np(pt, cName, sizeof(cName));
+            if (rc == 0) {
+                name = cName;
+            }
+        }
+    }
+}
+
+std::vector<ThreadSummary> EMGStackTraceRecorder::collectThreadSummaries() {
+    std::lock_guard<std::mutex> lockGuard(threadsLock);
+    
+    std::vector<ThreadSummary> summaries;
+    for (const auto &[threadId, thread] : threadsMap) {
+        std::vector<StackSummary> stackSummaries;
+        for (const auto &stack : thread.stacks) {
+            std::vector<uintptr_t> addresses;
+            for (auto i = stack.storageStartIndex; i < stack.storageEndIndex; i++) {
+                addresses.emplace_back(addressStorage[i]);
+            }
+            // Reverse the stack addresses to get the correct order
+            std::reverse(addresses.begin(), addresses.end());
+            stackSummaries.emplace_back(stack.time, addresses);
+        }
+        summaries.emplace_back(threadId, thread.name, stackSummaries);
+    }
+    return summaries;
+}
+
+void EMGStackTraceRecorder::recordStackForAllThreads(bool recordAllThreads, thread_t mainMachThread, thread_t etTraceThread) {
+    std::lock_guard<std::mutex> lockGuard(threadsLock);
```

Comment:
Did you also want to remove this by ensuring the collectThreadSummaries was always called on the same recording thread? Or is that not a bottleneck so not worth doing?

Reply:
It was less of a performance thing, more of a logic thing (e.g., not accidentally putting code above the lock in the future). However, doing it the other way has its own complications, so I just stuck with the mutex approach